### DISCUSSION TITLE
writing an app extension?

### DISCUSSION THREAD

**Discussion by rjkroege** [progrium/darwinkit#272] [Q&A]
# TL;DR
I wanted to write a *Safari Web Extension* (in Go obviously.) Like this [Swift example code from Apple](https://developer.apple.com/documentation/safariservices/developing-a-safari-web-extension?language=objc). Has anybody done this? How did you write the `main()` for the extension binary?

# Long Blither
I know how to convert the main Swift application to Go. But it's completely opaque to me how to create the `main()` for the *extension* process. Xcode tells me when breaking at the entry to the `NSExtensionRequestHandling` implementation (in the *extension*):

```
    frame #0: 0x0000000102611780 Sea Creator Extension.debug.dylib`SafariWebExtensionHandler.beginRequest(context=0x000000012b604870) at SafariWebExtensionHandler.swift:13:28
    frame #1: 0x0000000102612534 Sea Creator Extension.debug.dylib`@objc SafariWebExtensionHandler.beginRequest(with:) at <compiler-generated>:0
    frame #2: 0x00000001f302a038 ExtensionFoundation`__112-[EXConcreteExtensionContextVendor _beginRequestWithExtensionItems:listenerEndpoint:withContextUUID:completion:]_block_invoke.102 + 212
    frame #3: 0x00000001f301e274 ExtensionFoundation`-[EXExtensionContextImplementation _willPerformHostCallback:] + 92
    frame #4: 0x00000001f3029c90 ExtensionFoundation`__112-[EXConcreteExtensionContextVendor _beginRequestWithExtensionItems:listenerEndpoint:withContextUUID:completion:]_block_invoke + 928
    frame #5: 0x00000001959428f8 libdispatch.dylib`_dispatch_call_block_and_release + 32
    frame #6: 0x0000000195944658 libdispatch.dylib`_dispatch_client_callout + 20
    frame #7: 0x0000000195952f68 libdispatch.dylib`_dispatch_main_queue_drain + 980
    frame #8: 0x0000000195952b84 libdispatch.dylib`_dispatch_main_queue_callback_4CF + 44
    frame #9: 0x0000000195c1ce60 CoreFoundation`__CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
    frame #10: 0x0000000195bdca4c CoreFoundation`__CFRunLoopRun + 1996
    frame #11: 0x0000000195bdbbc4 CoreFoundation`CFRunLoopRunSpecific + 588
    frame #12: 0x0000000196d9a808 Foundation`-[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 212
    frame #13: 0x0000000196e12154 Foundation`-[NSRunLoop(NSRunLoop) run] + 64
    frame #14: 0x000000019580e718 libxpc.dylib`_xpc_objc_main + 700
    frame #15: 0x000000019581e824 libxpc.dylib`_xpc_main + 276
    frame #16: 0x000000019580e2b4 libxpc.dylib`xpc_main + 64
    frame #17: 0x0000000196db98e0 Foundation`-[NSXPCListener resume] + 308
    frame #18: 0x00000001a4db8534 PlugInKit`___lldb_unnamed_symbol1080 + 392
    frame #19: 0x00000001a4db83a4 PlugInKit`___lldb_unnamed_symbol1079 + 20
    frame #20: 0x00000001a4db8010 PlugInKit`___lldb_unnamed_symbol1073 + 552
    frame #21: 0x00000001a4db8904 PlugInKit`___lldb_unnamed_symbol1084 + 16
    frame #22: 0x00000001f3025188 ExtensionFoundation`EXExtensionMain + 304
    frame #23: 0x0000000196e1437c Foundation`NSExtensionMain + 204
    frame #24: 0x0000000195774274 dyld`start + 2840
```

I guess that I'd have to reverse engineer what's going on in `PluginKit` (frames [24-18]) to setup the XPC service (i.e. register the XPC service so that Safari can find it?) This service is responding to messages from Safari so I'd also have to write / call code to decode the incoming XPC messages (frames [4-2])? I'm questioning the effort-reward ratio here. :-)

**Comment by progrium** [comment]
I'm also curious to know! Would be a great example.
